File: src/benchmark.ts
import type {
  EngineAdapter,
  EngineBenchmarkResult,
  PromptRunResult,
} from "./types.js";
import { timedCompletion } from "./harness/measure.js";
import { DEFAULT_PROMPTS, WARMUP_PROMPT } from "./prompts.js";

export interface RunBenchmarkOptions {
  model: string;
  maxTokens?: number;
  serverStartTimeoutMs?: number;
  requestTimeoutMs?: number;
  prompts?: string[];
  verbose?: boolean;
  onProgress?: (line: string) => void;
}

const DEFAULT_MAX_TOKENS = 200;
const DEFAULT_SERVER_START_TIMEOUT_MS = 5 * 60 * 1000;
const DEFAULT_REQUEST_TIMEOUT_MS = 2 * 60 * 1000;

let nextPort = 41000;
function claimPort(): number {
  return nextPort++;
}

/**
 * Benchmarks a single engine. Never throws for an engine-level failure
 * (not installed, server start timeout, request failure) -- per the CEO
 * review's per-engine isolation decision, one bad engine never blocks
 * results from the others. Callers get a result object with installed:false
 * or per-run errors instead.
 */
export async function benchmarkEngine(
  adapter: EngineAdapter,
  opts: RunBenchmarkOptions,
): Promise<EngineBenchmarkResult> {
  const progress = opts.onProgress ?? (() => {});
  const prompts = opts.prompts ?? DEFAULT_PROMPTS;
  const maxTokens = opts.maxTokens ?? DEFAULT_MAX_TOKENS;

  if (!adapter.isInstalled()) {
    progress(`${adapter.name}: not installed, skipped`);
    return {
      engine: adapter.name,
      installed: false,
      error: `binary "${adapter.binary}" not found`,
      runs: [],
      avgTokensPerSecond: null,
      minTokensPerSecond: null,
      maxTokensPerSecond: null,
    };
  }

  progress(`${adapter.name}: starting server...`);
  let server;
  try {
    server = await adapter.startServer({
      model: opts.model,
      port: claimPort(),
      timeoutMs: opts.serverStartTimeoutMs ?? DEFAULT_SERVER_START_TIMEOUT_MS,
      verbose: opts.verbose,
    });
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    progress(`${adapter.name}: failed to start (${message})`);
    return {
      engine: adapter.name,
      installed: true,
      error: message,
      runs: [],
      avgTokensPerSecond: null,
      minTokensPerSecond: null,
      maxTokensPerSecond: null,
    };
  }

  try {
    progress(`${adapter.name}: warming up...`);
    try {
      await timedCompletion({
        engine: adapter.name,
        baseUrl: server.baseUrl,
        modelId: server.modelId,
        prompt: WARMUP_PROMPT,
        maxTokens: 16,
        timeoutMs: opts.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS,
      });
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      progress(`${adapter.name}: warm-up failed (${message})`);
      return {
        engine: adapter.name,
        installed: true,
        error: `warm-up failed: ${message}`,
        runs: [],
        avgTokensPerSecond: null,
        minTokensPerSecond: null,
        maxTokensPerSecond: null,
      };
    }

    const runs: PromptRunResult[] = [];
    for (let i = 0; i < prompts.length; i++) {
      const prompt = prompts[i];
      progress(`${adapter.name}: [${i + 1}/${prompts.length}] benchmarking...`);
      try {
        const result = await timedCompletion({
          engine: adapter.name,
          baseUrl: server.baseUrl,
          modelId: server.modelId,
          prompt,
          maxTokens,
          timeoutMs: opts.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS,
        });
        runs.push({ prompt, ok: true, result });
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        runs.push({ prompt, ok: false, error: message });
      }
    }

    const valid = runs
      .filter((r) => r.ok && r.result?.tokensPerSecond)
      .map((r) => r.result!.tokensPerSecond!);

    return {
      engine: adapter.name,
      installed: true,
      runs,
      avgTokensPerSecond:
        valid.length > 0
          ? Math.round((valid.reduce((a, b) => a + b, 0) / valid.length) * 100) / 100
          : null,
      minTokensPerSecond: valid.length > 0 ? Math.min(...valid) : null,
      maxTokensPerSecond: valid.length > 0 ? Math.max(...valid) : null,
    };
  } finally {
    server.stop();
  }
}

File: python/src/inferbench/benchmark.py
"""
Ported from src/benchmark.ts. Benchmarks a single engine. Never raises for
an engine-level failure (not installed, server start timeout, request
failure) -- engines are isolated from each other by design, so one bad
engine never blocks results from the others. Callers get a result object
with installed=False or per-run errors instead.
"""
from __future__ import annotations

from typing import Callable, List, Optional

from .harness.measure import TimedCompletionOptions, timed_completion
from .prompts import DEFAULT_PROMPTS, WARMUP_PROMPT
from .types import EngineAdapter, EngineBenchmarkResult, PromptRunResult

DEFAULT_MAX_TOKENS = 200
DEFAULT_SERVER_START_TIMEOUT_MS = 5 * 60 * 1000
DEFAULT_REQUEST_TIMEOUT_MS = 2 * 60 * 1000

_next_port = 41000


def _claim_port() -> int:
    global _next_port
    port = _next_port
    _next_port += 1
    return port


def benchmark_engine(
    adapter: EngineAdapter,
    *,
    model: str,
    max_tokens: Optional[int] = None,
    server_start_timeout_ms: Optional[int] = None,
    request_timeout_ms: Optional[int] = None,
    prompts: Optional[List[str]] = None,
    verbose: bool = False,
    on_progress: Optional[Callable[[str], None]] = None,
) -> EngineBenchmarkResult:
    progress = on_progress or (lambda _line: None)
    prompt_list = prompts if prompts is not None else DEFAULT_PROMPTS
    tokens_cap = max_tokens if max_tokens is not None else DEFAULT_MAX_TOKENS

    if not adapter.is_installed():
        progress(f"{adapter.name}: not installed, skipped")
        return EngineBenchmarkResult(
            engine=adapter.name,
            installed=False,
            error=f'binary "{adapter.binary}" not found',
        )

    progress(f"{adapter.name}: starting server...")
    try:
        server = adapter.start_server(
            model=model,
            port=_claim_port(),
            timeout_ms=server_start_timeout_ms or DEFAULT_SERVER_START_TIMEOUT_MS,
            verbose=verbose,
        )
    except Exception as err:  # noqa: BLE001 -- any start-up failure reports as a failed result, never raises
        message = str(err)
        progress(f"{adapter.name}: failed to start ({message})")
        return EngineBenchmarkResult(engine=adapter.name, installed=True, error=message)

    try:
        progress(f"{adapter.name}: warming up...")
        try:
            timed_completion(
                TimedCompletionOptions(
                    engine=adapter.name,
                    base_url=server.base_url,
                    model_id=server.model_id,
                    prompt=WARMUP_PROMPT,
                    max_tokens=16,
                    timeout_ms=request_timeout_ms or DEFAULT_REQUEST_TIMEOUT_MS,
                )
            )
        except Exception as err:  # noqa: BLE001 -- any warm-up failure reports as a failed result
            message = str(err)
            progress(f"{adapter.name}: warm-up failed ({message})")
            return EngineBenchmarkResult(
                engine=adapter.name,
                installed=True,
                error=f"warm-up failed: {message}",
            )

        runs: List[PromptRunResult] = []
        for i, prompt in enumerate(prompt_list):
            progress(f"{adapter.name}: [{i + 1}/{len(prompt_list)}] benchmarking...")
            try:
                result = timed_completion(
                    TimedCompletionOptions(
                        engine=adapter.name,
                        base_url=server.base_url,
                        model_id=server.model_id,
                        prompt=prompt,
                        max_tokens=tokens_cap,
                        timeout_ms=request_timeout_ms or DEFAULT_REQUEST_TIMEOUT_MS,
                    )
                )
                runs.append(PromptRunResult(prompt=prompt, ok=True, result=result))
            except Exception as err:  # noqa: BLE001 -- one failed prompt must not abort the sweep
                runs.append(PromptRunResult(prompt=prompt, ok=False, error=str(err)))

        valid = [
            r.result.tokens_per_second
            for r in runs
            if r.ok and r.result is not None and r.result.tokens_per_second
        ]

        return EngineBenchmarkResult(
            engine=adapter.name,
            installed=True,
            runs=runs,
            avg_tokens_per_second=round(sum(valid) / len(valid), 2) if valid else None,
            min_tokens_per_second=min(valid) if valid else None,
            max_tokens_per_second=max(valid) if valid else None,
        )
    finally:
        server.stop()
